### PR TITLE
Add design doc for proposed changes to `mz_internal`

### PR DESCRIPTION
**Pull request by umanwizard**
See rendered doc for details.

### CODE REVIEW

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+* `raw` variants should be removed, and the underlying dataflows
+  should be in terms of intervals.
+* All histogram views or sources should have a per-bucket `sum`
+  column.
+* All per-replica sources and views should have their names prepended
+  with some string (perhaps `clusterd_`, though I'm open to
+  suggestions) that makes it obvious which ones they are. Furthermore,
+  the limitations of per-replica relations should be clearly
+  documented.
+* The `mz_` prefix should be removed from all names.
```
Comment (benesch) [inline doc/developer/design/20230225_compute_introspection_schema.md:181]: I'm against this! Everything in `mz_catalog` has the `mz_` prefix to reduce the chance of a namespace collision with a user table. While `mz_internal` is not on the search path by default, many things in there are awaiting promotion to `mz_catalog` where they'll need the `mz_`, and I think we should just stick with the convention that everything in a `mz_` schema starts with an `mz_` prefix.

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  where the results are aggregated across workers in the way most
+  likely to be useful (for example, summing, taking the minimum, or
+  restricting to worker 0). Because these global views are more likely
+  to be useful than the underlying sources, they should have the basic
+  name (e.g., `mz_arrangement_sharing`). The underlying source should
+  have a name with `per_worker` inserted (e.g.,
+  `mz_per_worker_arrangement_sharing`). Furthermore, the per-worker
+  variant should be clearly less prominent in documentation (either at
+  the bottom of the page, or in a "see more" section that has to be
+  clicked to expand).
```
Comment (benesch) [inline doc/developer/design/20230225_compute_introspection_schema.md:164]: _Strong_ agree on this.

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+* The expectations about per-worker relations should be clearly
+  documented (e.g., we should clearly say whether the values for each
+  worker are expected to be the same during normal operation).
+* All relations that report a simple count or sum should be views based on
+  an `_internal` variant that maintains the count or sum in the
+  Differential diff field. We may consider renaming this, since it
+  conflicts confusingly with the overall name `mz_internal`.
+* `raw` variants should be removed, and the underlying dataflows
+  should be in terms of intervals.
```
Comment (benesch) [inline doc/developer/design/20230225_compute_introspection_schema.md:173]: @frankmcsherry was 👎🏽 on this in the past because the underlying data has nanosecond precision and our interval type only has microsecond precision.

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+* All histogram views or sources should have a per-bucket `sum`
+  column.
+* All per-replica sources and views should have their names prepended
+  with some string (perhaps `clusterd_`, though I'm open to
+  suggestions) that makes it obvious which ones they are. Furthermore,
+  the limitations of per-replica relations should be clearly
+  documented.
```
Comment (benesch) [inline doc/developer/design/20230225_compute_introspection_schema.md:180]: "clusterd" is not a user facing concept (outside of architecture discussion), but "cluster" is! Would strongly endorse calling these something with `cluster` in the name.

Cluster unification really did us a solid here, since we no longer need to figure out how to disambiguate between "storage" introspection views and "compute" introspection views.

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+The only way to understand the structure of scope nesting currently is
+to use the `mz_dataflow_addresses` relation and do some tedious list
+manipulation. A simple view of `(id, parent_id)` would make this much
+simpler.
+
+#### More views enriched with operator names, dataflow names, etc.
+
+(Not quite sure about this -- it's tricky to balance ease-of-use with
+not blowing up the set of documented relations)
+
```
Comment (benesch) [inline doc/developer/design/20230225_compute_introspection_schema.md:201]: I believe @frankmcsherry has a laundry list of controller-maintained introspection views that would be nice to have. Is this design document the right place to begin gathering those, or are those best left to a future design document?

**Review by benesch — COMMENTED** [review]
Really excited to see this work happening, @umanwizard!

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 4 above)
Comment (teskje) [reply doc/developer/design/20230225_compute_introspection_schema.md:180]: This is tough. I would have intuitively suggested that we prefix with `cluster_replica`, to signify that these relations are implicitly scoped by the targeted replica. However, we already have a bunch of `mz_cluster_replica_*` relations that are written by the controller and expose the global view (with a `replica_id` column):

* [mz_cluster_replica_metrics](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_cluster_replica_metrics)
* [mz_cluster_replica_sizes](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_cluster_replica_sizes)
* [mz_cluster_replica_statuses](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_cluster_replica_statuses)
* [mz_cluster_replica_utilization](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_cluster_replica_utilization)
* [mz_cluster_replica_frontiers](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_cluster_replica_frontiers)

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 5 above)
Comment (teskje) [inline doc/developer/design/20230225_compute_introspection_schema.md:201]: Are we up for rethinking the contents and structure of existing introspection sources? For example, I've been wondering if we really need to expose dataflow/channel addresses. If we are only interested in scopes, we could expose those more directly instead. And instead of exposing channels with their addresses, which then require some work to map them to the operators they connect, we could maybe expose channels with their operator IDs directly.

I think it would be a good idea to write down all the per-replica concepts we are interested in, their attributes and relations (e.g., as an ER diagram) and then translating that into introspection sources more or less directly. I appreciate that this is probably out-of-scope of this design doc.

**Review by teskje — COMMENTED** [review]
Thanks for writing this up!

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 4 above)
Comment (teskje) [reply doc/developer/design/20230225_compute_introspection_schema.md:180]: FWIF, here is some prior discussion about the `mz_cluster_replica_*` naming [in Slack](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1670937659099759).

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 3 above)
Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:173]: I think having two copies of every relation involving times is way more noise than it's worth. We can either have nanos everywhere as @teskje suggests, or we can give up nanosecond precision in the underlying sources. (How often is sub-microsecond precision useful, anyway?)

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 4 above)
Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:180]: @benesch , I think it's fine to expose non-user-facing concepts in internal stuff. I mean, dataflow channels and operators aren't a user-facing concept either :) 

I liked `clusterd` because it precisely describes where the logging happens and where the relations actually live (as opposed to being persisted). I'd also be okay with `cluster_replica` (and renaming all the stuff that already has it)

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+## Non-Goals
+
+* The relations do not need to be polished enough for external
+consumption (although it's hoped they prove useful, we don't consider
+it a bug if they're not). Customer-friendly introspection will be
+provided in the web UI by the "external introspection" project.
+
+* We make no attempt to maintain backwards compatibility with
+  previously-existing queries, although as part of this work we will
+  update the internal tools that use the introspection relations.
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:28]: One of the prometheus SQL exporters probably does. Also the CPU/Memory utilization graphs in the web UI. We will need to fix these if we rename things.

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+## Description
+
+### Schema inconsistencies and inelegancies 
+
+In this section, we describe the various issues that make the schema
+confusing.
+
+#### Worker-local vs. global
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:37]: Good point, it is usually per-replica. I'll change this heading to "Per-worker vs. aggregated".

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+`mz_compute_frontiers`. "Sharded" might not be the most precise way to
+describe these, but I include them in this section because the point
+is basically the same.
+
+The existence of the per-worker underlying sources is not completely
+useless, e.g. because they are useful for detecting skew. However,
+they are probably less commonly useful than a summed view would
+be. Unforunately, this summed view does not exist for most such
+relations (one counterexample is `mz_records_per_dataflow` and
+`mz_records_per_dataflow_global`).
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:66]: Yes, however this will be fixed by the other suggestions in this document (renaming `mz_records_per_dataflow_global` to `mz_records_per_dataflow` and `mz_records_per_dataflow` to `mz_records_per_dataflow_per_worker`)

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 5 above)
Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:201]: We can certainly discuss this here. I think the main things we care about are:

* Channels (and the number/size of messages sent on them)
* Operators (and the number/size of objects they retain, which is only relevant for arrangements)
* Scopes and their parent/child relationships
* Dataflows and the dependencies among them
* Frontiers

I agree that I can't immediately think of a good reason to expose addresses.

Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:201]: One thing Frank mentioned on Slack:

> In particular, it is hard to get end-to-end lag (from sources all the way trhough to indexes) though "easy" to get lag for one dataflow (it's inputs to outputs).

We should definitely have this, if we think of a good way to actually compute it ... you need to know the transitive dependencies of each dataflow, which is fundamentally recursive, and we don't (yet) have recursive queries outside of unsafe mode. I'll check whether it's easy to get transitive dependencies in the code.

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+  variant should be clearly less prominent in documentation (either at
+  the bottom of the page, or in a "see more" section that has to be
+  clicked to expand).
+* The expectations about per-worker relations should be clearly
+  documented (e.g., we should clearly say whether the values for each
+  worker are expected to be the same during normal operation).
+* All relations that report a simple count or sum should be views based on
+  an `_internal` variant that maintains the count or sum in the
+  Differential diff field. We may consider renaming this, since it
+  conflicts confusingly with the overall name `mz_internal`.
```
[the remark this answers is not in the document]
Comment (teskje) [reply doc/developer/design/20230225_compute_introspection_schema.md:171]: I think it would mostly be adding a flag somewhere and using that to decide whether to write an introspection source to `mz_sources` or not. But I agree on this not being high-priority.

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 5 above)
Comment (teskje) [reply doc/developer/design/20230225_compute_introspection_schema.md:201]: Here is what I've come up with as a first draft:
Blank diagram

It's missing a couple thing (parks, delays), but most of what we currently have is there. The things marked with * are changing as part of dataflow processing. We should probably separate these from things that only change as part of dataflow creation/dropping, to minimize the size of changes. I think if you do that you end up somewhat close to what we already have, which is reassuring.

The most debatable thing about the approach I depicted is probably that grouping operators inside scopes requires you to use `WITH MUTUALLY RECURSIVE` to enumerate all the operators inside one dataflow, which will be great in half a year, but not today. We should probably connect the operators to their dataflow directly too.

Also, our current introspection sources mirror the structure of timely's/DD's logging quite closely. Having this more user-centric view will probably make the logging logic in clusterd more of a hassle to implement.

Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:201]: Can you summarize the main differences between this proposal and what we have now?

Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:201]: My intuition is that this will end up being a rather large change; we should probably discuss it as separate from this design doc (as you suggested initially)

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 4 above)
Comment (teskje) [reply doc/developer/design/20230225_compute_introspection_schema.md:178]: I would be great to find something else than `cluster_replica`. `mz_cluster_replicas` is already a stable system table that we cannot rename, so using the very similar `mz_cluster_replica` as an introspection source prefix would be confusing.

Some brainstorm ideas, none of them great:
* `clusterd`. Already mentioned above.
* `cluster_side`. To avoid the internal name.
* `replica_side`. A bit more specific/correct.
* `replica`. Maybe harder to confuse than `cluster_replica`, but still quite confusing. It's not obvious that leaving out the "cluster" would have significance.
* `compute`. We will use the introspection sources for storage dataflows too, so this would not be referring to COMPUTE the layer, but to our computation fabric.
* `computation`. Less loaded alternative to the above.
* Using a prefix different then `mz_`, strawman: `log_`.
* Using a schema different from `mz_internal`, strawman: `mz_log`.

**Comment by umanwizard** [comment]
I'm now thinking about whether we should also expose per-operator frontiers. I'm not sure why we don't have them; maybe the data volume was thought to be too high?

**Comment by teskje** [comment]
> I'm now thinking about whether we should also expose per-operator frontiers. I'm not sure why we don't have them; maybe the data volume was thought to be too high?

We have `mz_dataflow_operator_reachability`, which I think is basically the same as operator frontiers? We don't have documentation for it, so I'm not entirely sure, but at least the amount of data should be similar to that of operator frontier tracking.

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+(Not quite sure about this -- it's tricky to balance ease-of-use with
+not blowing up the set of documented relations)
+
+## Alternatives
+
+None known to me
+
+## Open questions
+
+* Are there further views that could be useful but that we don't have
```
Comment (vmarcos) [inline doc/developer/design/20230225_compute_introspection_schema.md:213]: When calculating frontier lag, we usually have to write some tedious joins between `mz_compute_frontiers` and `mz_compute_import_frontiers`. Perhaps a ready-made view there would be pretty useful?

**Review by vmarcos — APPROVED** [review]
Thanks for taking this on! Overall, the suggestions here are really great and will help us a lot.

**Comment by vmarcos** [comment]
Ah, one other comment that crossed my mind: Our treatment of `SUBSCRIBE`s in introspection sources / views could probably be improved. For example, in `mz_compute_frontiers` et al., we show `SUBSCRIBE`s, though the documentation is inconsistent about it. Also, I am not sure whether we have a list of the `SUBSCRIBE`s that are running along with their SQL statements? We can look at dataflows and do some string matching, which is not ideal (but that is what our documentation suggests [here](https://materialize.com/docs/ops/troubleshooting/#how-many-subscribe-commands-are-running)).

**Comment by teskje** [comment]
> Our treatment of SUBSCRIBEs in introspection sources / views could probably be improved.

Maybe we should just add an `mz_subscribes` table? Subscribes don't have names, but we could put in (time, user, SQL) instead, to make them identifiable.

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+  documented (e.g., we should clearly say whether the values for each
+  worker are expected to be the same during normal operation).
+* All relations that report a simple count or sum should be views based on
+  an `_internal` variant that maintains the count or sum in the
+  Differential diff field. We may consider renaming this, since it
+  conflicts confusingly with the overall name `mz_internal`.
+* `raw` variants should be removed, and the underlying dataflows
+  should be in terms of intervals.
+* All histogram views or sources should have a per-bucket `sum`
+  column.
```
Comment (teskje) [inline doc/developer/design/20230225_compute_introspection_schema.md:171]: Does that imply that we won't be able to use DD's diff field anymore for maintaining the bucket counts? I think that roughly doubles our traffic on those collections, because we now need to retract previous values, which we didn't have to before. Probably still fine, since the number of buckets limits how many changes we can see per time interval.

Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:171]: The DD diff field can be a two-element `(sum, count)` tuple, so in principle you could still use it. When I implemented bucket counts for `frontier_delay` we went back and forth on this a few times.

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+##### Sharded relations
+
+*Example:* `mz_arrangement_sizes`. These are relations that report
+some value that is expected to differ across all workers, such that
+the semantically meaningful value is (often) the sum of all the
+per-worker values. For example, a user interested in the total size of
+each arrangement would normally want to issue a query like `SELECT
+operator_id, sum(records), sum(batches) FROM mz_arrangement_sizes
+GROUP BY worker_id`.
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:56]: Yes, thanks

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 11 above)
Comment (teskje) [reply doc/developer/design/20230225_compute_introspection_schema.md:171]: But when we want to sink to persist, our diff type can only be `i64`, right?

Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:171]: You're right, it has to go to `i64` at any boundary with the outside world. I think it's okay, the traffic in these tables should be rather small.

Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:171]: s/tables/sources/

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+
+We should add the transitive closure of `mz_object_dependencies`,
+called e.g. `mz_transitive_object_dependencies`.
+
+This will allow us to see the end-to-end frontier lag for any
+dataflow.
+
+#### More views enriched with operator names, dataflow names, etc.
+
+(Not quite sure about this -- it's tricky to balance ease-of-use with
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:204]: Indeed! I plan to continue updating this PR if we discover any possible new views through that effort.

**Comment by teskje** [comment]
An `mz_subscriptions` table listing all active subscribes exists now: #17906.

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+  documented (e.g., we should clearly say whether the values for each
+  worker are expected to be the same during normal operation).
+* All relations that report a simple count or sum should be views based on
+  an `_internal` variant that maintains the count or sum in the
+  Differential diff field. We may consider renaming this, since it
+  conflicts confusingly with the overall name `mz_internal`.
+* `raw` variants should be removed, and the underlying dataflows
+  should be in terms of intervals.
+* All histogram views or sources should have a per-bucket `sum`
+  column.
```
[the remark this answers is not in the document]
Comment (teskje) [reply doc/developer/design/20230225_compute_introspection_schema.md:171]: We discussed offline the option to instead drop the `sum` field from all histograms. This is a summary of that discussion and a bit of research on my side.

* We introduced the `sum` field [because Prometheus expects it](https://prometheus.io/docs/practices/histograms/#count-and-sum-of-observations).
* The use of this field is that you can compute averages (both per bucket and across buckets).
* We can compute approximate the sum using the bucket upper-bound `b`:
    * `b * count` is an over-approximation. For our power-of-2 based histograms it is at worst off by a factor of 2.
    *  If we now the upper bound of the previous bucket `b'`, we could get a better approximation with `((b + b') / 2) * count`
* Setting the sum to 0 or leaving it out when reporting a histogram to Prometheus will make the resulting metric show 0. While it doesn't break anything, it will produce confusing results if one tries to compute an average.

So there are two viable options we have to decide between: Adding a sum field (or multiple) to each histogram source, or not doing that and approximating the sum when we need it.

I think I prefer the second option (approximating the sum) because:
* We don't know if we (or our users) need accurate averages.
* Not having the sum field means having less code, and possibly less resource usage for maintaining it.
* We can always add the field later if we find a strong use-case. Removing it later would be more disruptive (even for our unstable introspection sources).

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 3 above)
Comment (teskje) [reply doc/developer/design/20230225_compute_introspection_schema.md:169]: I discussed this with @umanwizard offline and we agree that keeping just the nanoseconds is fine. Users can still convert those to intervals themselves if they are fine with the loss in precision.

Based on brief testing, it seems like all duration values we currently produce are above 1000 ns. That said, there is no reason why this always has to be the case in the future. One might hope, for example, that our operators one day are fast enough to have sub-microsecond invocation times. So keeping the nanosecond-precision values is future-proof.

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:171]: I agree with @teskje here. The `sum` field is not so important that we need to keep it if it is introducing difficulties.

Note that we will have to modify the prometheus-sql-exporter in the cloud repo to continue understanding the histograms if we remove the sum field

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 4 above)
Comment (teskje) [reply doc/developer/design/20230225_compute_introspection_schema.md:193]: My most recent thinking is that we should prefix these sources/views with `mz_log`. We already present introspection sources as having type "log" in `mz_sources`, so this is kind of already a user-facing concept. I'm sure we can do a better job of explaining it in the documentation too.

I prefer "log" over "introspection" because it is significantly shorter. Not sure how good of an argument that is, but typing queries against introspection sources is already annoying because you have to prefix all of them with `mz_internal.` every time.

Comment (umanwizard) [reply doc/developer/design/20230225_compute_introspection_schema.md:193]: > We already present introspection sources as having type "log" in mz_sources, so this is kind of already a user-facing concept

Unfortunately, it is not quite the same concept. A relation that differs across a replica might be a log _or_ a view (which depends on logs). (Aside: another confusing point is that the terms "logs" and "sources" are often used interchangeably in this context (whereas "sources" from Storage's perspective are a rather different thing)).

Furthermore, logs might someday be queryable across replicas, if we re-introduce persisted introspection.

For all these reasons, I think object type (log vs. view) on one hand, and per-replica vs. global on the other hand, are two separate (but related) concepts.

One idea I had was to simply refuse to allow replica-directed queries without special syntax, e.g. something like `SELECT IN REPLICA (default.r1) * FROM mz_internal.mz_dataflow_channel_operators`. This would also allow us to get rid of the `cluster_replica` session variable.

Another reason why I like this idea is that replica-directed queries are one of the very few ways of introducing non-determinism in Materialize (the only other one I can think of is non-materializable functions). I.e., if the state of persist and catalogs, the selected timestamps, and the query issued are all held constant, the results are fully determined. However, replica-directed queries are an exception to this rule so it makes sense to force them to be queried with different syntax.

I would be interested in what @benesch or some of the SQL-focused surfaces folks (@mjibson , @jkosh44) think of this idea

Comment (jkosh44) [reply doc/developer/design/20230225_compute_introspection_schema.md:193]: I personally like the idea of having to specify a replica for replica specific queries. It makes it explicit what replica the query is going to, instead of having to remember the value of some session variable, and it makes it explicit that the query is specific to a single replica. I don't have a strong opinion on what the specific syntax should be, but I vaguely remember a previous conversation about where the `IN REPLICA` syntax needs to be in the query to maintain SQL compatibility.

Comment (teskje) [reply doc/developer/design/20230225_compute_introspection_schema.md:193]: I like the idea of having special query syntax too. I think the reason we didn't really consider it before was that people pointed out that changing the `SELECT` syntax is scary. I don't remember the details though. I think @sploiselle might?

Another thing to consider is that `SUBSCRIBE`s can also target replicas, so if we introduce dedicated syntax, that needs to apply to subscribes as well.

For me, the value of having a specific prefix for log sources and views over them is not so much that you can immediately see that they are replica-specific, but rather that they were collected on the replica-side, rather than on the environmentd-side. That's a subtle difference, since the latter applies even if we lift the replica-targeting restriction in the future. My thought was that we might want to collect the same information on both sides and need a way to express the distinction to the user. For example today we have `mz_cluster_replica_frontiers` and `mz_compute_frontiers`, and it's not really clear how they differ. That said, I'm not actually sure `mz_compute_frontiers` actually provides much additional value over `mz_cluster_replica_frontiers`, so maybe we should just remove it and resolve the ambiguity that way? If it is never the case that we want to collect the same information from both the replicas and environmentd, then a per-replica prefix for relation names is not needed either.

> Aside: another confusing point is that the terms "logs" and "sources" are often used interchangeably in this context (whereas "sources" from Storage's perspective are a rather different thing)

I've thought that maybe we should have an `mz_logs` table that's different from `mz_sources`. But I think right now the distinction between sources and tables is pretty muddled anyway, so there is opportunity for coming up with a holistic design separately.

**Comment by teskje** [comment]
> We have mz_dataflow_operator_reachability, which I think is basically the same as operator frontiers? We don't have documentation for it, so I'm not entirely sure, but at least the amount of data should be similar to that of operator frontier tracking.

I've checked this: `mz_dataflow_operator_reachability` doesn't give us operator frontiers directly. It gives us the capabilities held by each operator output and input. Frontiers can be derived from that, if you know the structure of the dataflow graph and the path summaries between individual operators. But a) we don't know the path summaries and b) we would surely not want to perform the work of deriving frontiers ourselves, even if we had all required information.

I think timely's reachability tracker could be made to log frontier updates as well as capability updates. If we want to show operator frontiers to users, this is the only feasible way I can think of.

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 4 above)
Comment (benesch) [reply doc/developer/design/20230225_compute_introspection_schema.md:193]: > @benesch , I think it's fine to expose non-user-facing concepts in internal stuff. I mean, dataflow channels and operators aren't a user-facing concept either :)

Dataflows, channels, and operators _are_ user facing, in my view. Only for the most advanced of users, but they are concepts that must be understood to make sense of the introspection relations.

By contrast, I would like for `clusterd` to never leak into the user interface at all. If `storaged` or `computed` had leaked into the interface, for example, that would have sank cluster unification.

> One idea I had was to simply refuse to allow replica-directed queries without special syntax, e.g. something like `SELECT IN REPLICA (default.r1) * FROM mz_internal.mz_dataflow_channel_operators`. This would also allow us to get rid of the `cluster_replica` session variable.
>
> Another reason why I like this idea is that replica-directed queries are one of the very few ways of introducing non-determinism in Materialize (the only other one I can think of is non-materializable functions). I.e., if the state of persist and catalogs, the selected timestamps, and the query issued are all held constant, the results are fully determined. However, replica-directed queries are an exception to this rule so it makes sense to force them to be queried with different syntax.

I wrote up thoughts about this in the past: https://github.com/MaterializeInc/materialize/issues/15195#issuecomment-1269229993

IMO, introducing special syntax _specifically_ for replica-targeted queries would be a mistake. There are plenty of other session variables that quietly impact how your query works (e.g., `database`, `search_path`, `cluster`). I don't see why `cluster_replica` is more special those session variables. The argument that it introduces non-determinism doesn't resonate for me, given that (as you say) `SELECT mz_version()` is _also_ nondeterministic, and we don't make you use special syntax to type that.

As mentioned in the comment I linked above, no objection to extending _all_ SQL statements to support setting arbitrary session variables.

> Not sure how good of an argument that is, but typing queries against introspection sources is already annoying because you have to prefix all of them with `mz_internal.` every time.

As of recently, you can say `set search_path = mz_catalog, mz_internal, public`, and you'll no longer need to prefix all `mz_internal` tables with the schema name.

> > Aside: another confusing point is that the terms "logs" and "sources" are often used interchangeably in this context (whereas "sources" from Storage's perspective are a rather different thing)
> 
> I've thought that maybe we should have an `mz_logs` table that's different from `mz_sources`. But I think right now the distinction between sources and tables is pretty muddled anyway, so there is opportunity for coming up with a holistic design separately.

Totally agree that we should separate the concepts of "logs" and "sources". 

I've never liked the term "log" though. It gets confused with what most people think of as a log: lines of text written to stdout and stderr. As unwieldy as "introspection" is, it's free of baggage.

Comment (benesch) [reply doc/developer/design/20230225_compute_introspection_schema.md:193]: The best idea I have is a prefix or suffix like `active_replica`. For example: `mz_active_replica_dataflow_operators`. It's wordy, though...

To come clean: pre-cluster unification I was opposed to ever using the term "replica" alone (vs. "cluster replica"). I've softened my stance there, since we no longer are thinking we'll need to distinguish between "storage replicas" and "cluster replicas".

File: doc/developer/design/20230225_compute_introspection_schema.md [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+  variant should be clearly less prominent in documentation (either at
+  the bottom of the page, or in a "see more" section that has to be
+  clicked to expand).
+* The expectations about per-worker relations should be clearly
+  documented (e.g., we should clearly say whether the values for each
+  worker are expected to be the same during normal operation).
+* All relations that report a simple count or sum should be views based on
+  an `_internal` variant that maintains the count or sum in the
+  Differential diff field. We may consider renaming this, since it
+  conflicts confusingly with the overall name `mz_internal`.
```
Comment (teskje) [inline doc/developer/design/20230225_compute_introspection_schema.md:175]: How about using a `_raw` suffix instead of `_internal`? That is free of ambiguity once we remove the `mz_raw_*` variants. Also reflects quite well that the data in these collections consists of the raw un-aggregated DD updates.

File: doc/developer/design/20230225_compute_introspection_schema.md (hunk 4 above)
Comment (teskje) [reply doc/developer/design/20230225_compute_introspection_schema.md:193]: > By contrast, I would like for clusterd to never leak into the user interface at all. If storaged or computed had leaked into the interface, for example, that would have sank cluster unification.

Assuming that "user interface" here only concerns the introspection sources, why would that have been an issue? Backwards compatibility shouldn't be a concern, since we declare all the introspection sources and views as unstable.

> As unwieldy as "introspection" is, it's free of baggage.

Although we now have the `mz_introspection` cluster, which uses the term for "running SHOW commands and the like". Which also means that adding a separate `mz_introspection` schema for introspection sources and views to live in would be confusing.

Maybe there is no need to assign a specific prefix to introspection sources. AFAICS, there are two use-cases for a special prefix:
1. Telling users which relations are per-replica and which are not. I think we can solve this with documentation, like having a separate section on the `mz_internal` catalog docs page that lists all the introspection sources. And hopefully we will end up with a solution that removes per-replica relations at some point.
2. Differentiating between sources that provide the same/similar data but collected at different places, for example `mz_compute_frontiers` (collected on the replica) vs. `mz_cluster_replica_frontiers` (collected by the controller). I think in many of these cases, we can just get rid of one of the two relations. If there is a controller-exported relation exposing the same data as a replica-exported one, the replica-exported one is probably not necessary. There will of course be exceptions to this rule, but maybe those can be solved on a case-by-case basis.
